### PR TITLE
CLICKHOUSE-784: reset merge tree setting

### LINKED ISSUE #14449: Allow to undo ALTER TABLE ... MODIFY SETTING
I run

```sql
ALTER TABLE target_table on CLUSTER '{cluster}' MODIFY SETTING merge_with_ttl_timeout = 3600
```

But there's no way to undo it. I want the default TTL value to be mirrored for the table so I need to reset this settings.

I can run

```sql
ALTER TABLE target_table on CLUSTER '{cluster}' MODIFY SETTING merge_with_ttl_timeout = 86400
```

but that still would be only static setting and default would not be mirrored.

### PR DESCRIPTION
**Pull request by xjewer**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add an ability to reset custom setting to default and remove it from table's metadata.
This will allow to rollback the change without knowing the system/config's default. Closes 
#14449. 

Example:

Modify setting (currently exists)
```SQL
ALTER TABLE foo MODIFY SETTING merge_with_ttl_timeout=1000;
```

Reset setting to config's default and remove from table metadata
```SQL
ALTER TABLE foo RESET SETTING merge_with_ttl_timeout;
```

### CODE REVIEW

File: src/Storages/AlterCommands.cpp [hunk 1]
```diff
@@ -524,6 +540,20 @@ void AlterCommand::apply(StorageInMemoryMetadata & metadata, const Context & con
...
+        auto & settings_from_storage = metadata.settings_changes->as<ASTSetQuery &>().changes;
+        for (const auto & setting_name : settings_resets)
+        {
+            auto finder = [&setting_name](const SettingChange & c) { return c.name == setting_name; };
+            auto it = std::find_if(settings_from_storage.begin(), settings_from_storage.end(), finder);
+
+            if (it != settings_from_storage.end())
+                settings_from_storage.erase(it);
+            else
+                throw Exception("Setting " + backQuote(setting_name) + " is not defined for a table", ErrorCodes::UNKNOWN_SETTING);
```
Comment (xjewer) [inline src/Storages/AlterCommands.cpp:554]: current implementation implies the RESET SETTING is not idempotent.

Was thinking what is the best approach here ignore a missing setting or add `IF EXISTS` statement?

Comment (filimonov) [reply src/Storages/AlterCommands.cpp:554]: Discussed in telegram. You should be able to always reset setting to default. Even if it was never adjusted for that table or was already reset before. 

Exception only for unknown / misspelled settings.

Comment (xjewer) [reply src/Storages/AlterCommands.cpp:554]: the only problem with this approach, that there is no understanding of the table engine in AlterCommand. Need to pass a set of settings changes with additional information (setting has to be removed) to storage.

`RESET SETTING IF EXISTS` would be easier to implement.

**Comment by alexey-milovidov** [comment]
> Was thinking what is the best approach here ignore a missing setting or add IF EXISTS statement?

Ignore if setting was not set is Ok.

**Comment by alexey-milovidov** [comment]
@xjewer Let's revive it.

**Comment by xjewer** [comment]
>Ignore if setting was not set is Ok.

it will mask errors, i.e. if setting name has a typo. Though it will be the same behaviour with `IF EXISTS` statement.

**Comment by xjewer** [comment]
@alexey-milovidov failed tests don't seem to be related to the change

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Review by alexey-milovidov — APPROVED** [review]

**Comment by xjewer** [comment]
failing 01781_merge_tree_deduplication test is suspicious

locally it passes:
```
Running 1 stateless tests (MainProcess).

01781_merge_tree_deduplication:                                         [ OK ]

1 tests passed. 0 tests skipped. 1.51 s elapsed (MainProcess).
Won't run stateful tests because test data wasn't loaded.
All tests have finished.
```

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-11546.

**Pull request merged**